### PR TITLE
tsify Grid

### PR DESCRIPTION
**Pull request by amanmahajan7**

### CODE REVIEW

File: packages/react-data-grid/src/Grid.tsx [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+  scrollLeft?: number;
+  onViewportKeydown(e: React.KeyboardEvent<HTMLDivElement>): void;
+  onViewportKeyup(e: React.KeyboardEvent<HTMLDivElement>): void;
+  onColumnResize(index: number, width: number): void;
+  onSort(columnKey: string, direction: DEFINE_SORT): void;
+  getValidFilterValues?(): unknown;
+  onHeaderDrop?(): void;
+}
+
+export default class Grid extends React.Component<Props> {
```
Comment (MayhemYDG) [inline packages/react-data-grid/src/Grid.tsx]: Looks like it's safe and easy to turn this one into a functional component.
We can replace the refs with `useRef`, and `_scrollLeft` with... `useRef`.
https://reactjs.org/docs/hooks-faq.html#is-there-something-like-instance-variables

File: packages/react-data-grid/src/Grid.tsx [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  static displayName = 'Grid';
+
+  private header = React.createRef<Header>();
+  private viewport = React.createRef<Viewport>();
```
Comment (MayhemYDG) [inline packages/react-data-grid/src/Grid.tsx]: If you want to keep this a class on the other hand, let's change these two to `readonly`.

File: packages/react-data-grid/src/Grid.tsx (hunk 1 above)
Comment (amanmahajan7) [reply packages/react-data-grid/src/Grid.tsx]: Let's revisit this after we fix the scroll logic

File: packages/react-data-grid/src/Grid.tsx (hunk 2 above)
Comment (amanmahajan7) [reply packages/react-data-grid/src/Grid.tsx]: Done

File: packages/react-data-grid/src/HeaderCell.tsx [hunk 3]
```diff
@@ -99,12 +99,12 @@ export default class HeaderCell extends React.Component<Props> {
...
 
-    if (rowType === HeaderRowType.HEADER && column.draggable) {
-      const DraggableHeaderCell = this.props.draggableHeaderCell;
+    const DraggableHeaderCell = this.props.draggableHeaderCell;
+    if (rowType === HeaderRowType.HEADER && column.draggable && DraggableHeaderCell) {
       return (
         <DraggableHeaderCell
           column={column}
-          onHeaderDrop={this.props.onHeaderDrop}
+          onHeaderDrop={this.props.onHeaderDrop!}
```
Comment (amanmahajan7) [inline packages/react-data-grid/src/HeaderCell.tsx]: We need a better way to validate `onHeaderDrop` prop. May be we can make `onHeaderProp` optional in `DraggableHeaderCell` but drag/drop does not work without it. I am not a fan of using a wrapper for adding draggable support 

https://github.com/adazzle/react-data-grid/blob/next/examples/scripts/example24-draggable-header.js#L80

File: packages/react-data-grid/src/Header.tsx [hunk 4]
```diff
@@ -15,16 +15,15 @@ interface HeaderRowData extends Pick<HeaderRowProps, 'filterable' | 'onFilterCha
 }
 
 type SharedHeaderRowProps = Pick<HeaderRowProps,
-'draggableHeaderCell'
-| 'getValidFilterValues'
-| 'sortDirection'
-| 'sortColumn'
-| 'onHeaderDrop'
-| 'onSort'
-| 'onScroll'
```
Comment (amanmahajan7) [inline packages/react-data-grid/src/Header.tsx]: Grid never provides `onScroll` prop
https://github.com/adazzle/react-data-grid/pull/1614/files#diff-8c020d9d11e86efdcd2e7b42efd9fa29L139

File: packages/react-data-grid/src/Header.tsx [hunk 5]
```diff
@@ -15,12 +15,12 @@ interface HeaderRowData extends Pick<HeaderRowProps, 'filterable' | 'onFilterCha
...
-  | 'sortDirection'
-  | 'sortColumn'
-  | 'onHeaderDrop'
-  | 'onSort'
+'draggableHeaderCell'
+| 'getValidFilterValues'
+| 'sortDirection'
+| 'sortColumn'
+| 'onHeaderDrop'
+| 'onSort'
```
Comment (MayhemYDG) [inline packages/react-data-grid/src/Header.tsx]: It'll be fixed eventually, gotta give it time.
https://github.com/typescript-eslint/typescript-eslint/issues/455

**Pull request merged**